server/src/main/java/org/candlepin/controller/ContentAccessManager.java

```diff
@@ -559,6 +543,32 @@ else if (pc.isEnabled() && !contentEnabledMap.get(pc.getContent().getUuid())) {
             emptyConsumer, emptyPool, null);
     }
 
+    /**
+     * To process product content and added it to container product.
+     *
+     * @param container
+     *  Product container
+     *
+     * @param ownerId
+     *  Id of an owner
+     *
+     */
+    private void processOwnerContent(Product container, String ownerId) {
```

Comment:
Ah ha. I remember this discussion now.

So we need to include all "active" content for a given org, but we need to flag it as enabled properly, resolving any conflicts to enabled.

That complicates things a bit, but is something we can still solve efficiently.

Since ProductContent is a join object, we don't want to fetch it directly if we can help it -- especially since the object we *actually* care about is the Content instance, which we have to fetch even if we try to use the ProductContent mutators of Product directly. Using ProductContent like this results in another N+1 issue, where examining the content field results in another query. Running the spec test for this method, for instance, results in the following queries:

```
select pool0_.id as col_0_0_ from cp_pool pool0_ where pool0_.owner_id=? and pool0_.startDate<=? and pool0_.endDate>=?

SELECT pool.product_uuid FROM cp_pool pool WHERE pool.id IN (?, ?) UNION SELECT pool.derived_product_uuid FROM cp_pool pool WHERE pool.id IN (?, ?) UNION SELECT pp.product_uuid FROM cp2_pool_provided_products pp WHERE pp.pool_id IN (?, ?) UNION SELECT dpp.product_uuid FROM cp2_pool_derprov_products dpp WHERE dpp.pool_id IN (?, ?) 

select productcon0_.id as id1_12_, productcon0_.created as created2_12_, productcon0_.updated as updated3_12_, productcon0_.content_uuid as content_5_12_, productcon0_.enabled as enabled4_12_, productcon0_.product_uuid as product_6_12_ from cp2_product_content productcon0_ where productcon0_.product_uuid in (? , ? , ?)

select content0_.uuid as uuid1_1_0_, content0_.created as created2_1_0_, content0_.updated as updated3_1_0_, content0_.arches as arches4_1_0_, content0_.contentUrl as contentU5_1_0_, content0_.entity_version as entity_v6_1_0_, content0_.gpgUrl as gpgUrl7_1_0_, content0_.content_id as content_8_1_0_, content0_.label as label9_1_0_, content0_.locked as locked10_1_0_, content0_.metadataExpire as metadat11_1_0_, content0_.name as name12_1_0_, content0_.releaseVer as release13_1_0_, content0_.requiredTags as require14_1_0_, content0_.type as type15_1_0_, content0_.vendor as vendor16_1_0_ from cp2_content content0_ where content0_.uuid=?

select content0_.uuid as uuid1_1_0_, content0_.created as created2_1_0_, content0_.updated as updated3_1_0_, content0_.arches as arches4_1_0_, content0_.contentUrl as contentU5_1_0_, content0_.entity_version as entity_v6_1_0_, content0_.gpgUrl as gpgUrl7_1_0_, content0_.content_id as content_8_1_0_, content0_.label as label9_1_0_, content0_.locked as locked10_1_0_, content0_.metadataExpire as metadat11_1_0_, content0_.name as name12_1_0_, content0_.releaseVer as release13_1_0_, content0_.requiredTags as require14_1_0_, content0_.type as type15_1_0_, content0_.vendor as vendor16_1_0_ from cp2_content content0_ where content0_.uuid=?
```

Note the two content queries at the end -- that's invoked once for each ProductContent fetched by the getProductContentsOfActivePoolsByOwner\* method when getContent is invoked. Side note: that method name is a bit wordy -- getActiveProductContentByOwner would suffice, and even that's a tad long due to specifying the return type in the method name.

So, that means we need to solve this in the curator method.

My first thought was to use a DISTINCT with ORDER BY on the enabled field in ascending order, and then collect the content objects in order. This would put all the enabled content at the end, and would overwrite any disabled entries with an enabled one:

```
jpql = "SELECT DISTINCT pc.content, pc.enabled FROM ProductContent pc WHERE pc.product.uuid IN (:product_uuids) ORDER BY pc.enabled ASC";
```

This would work perfectly if it wasn't for one small hiccup: since we're partitioning, we have to assume we'll be running this query at least twice; and if we're running it twice, our order is no longer guaranteed between queries, and we run the risk of clobbering our "enabled" entries from the end of previous query with "disabled" entries in the next. We can work around this with some logic similar to what is outlined in the method this comment hangs on, but at that point, the ordering does literally nothing for us -- so we can omit it entirely.

```
    // OwnerContentCurator.java
    public Map<Content, Boolean> getActiveContentByOwner(String ownerId) {
        EntityManager entityManager = this.getEntityManager();
        Date now = new Date();

        // Get all the active pools

        String jpql = "SELECT p.id FROM Pool p " +
            "WHERE p.owner.id = :owner_id AND p.startDate <= :start_date AND p.endDate >= :end_date";

        List<String> poolIds = entityManager.createQuery(jpql, String.class)
            .setParameter("owner_id", ownerId)
            .setParameter("start_date", now)
            .setParameter("end_date", now)
            .getResultList();

        // Use the pool IDs to select all related product UUIDs
        // Note: we have to start partitioning now to safely handle any number of active pools

        String sql = "SELECT pool.product_uuid FROM cp_pool pool WHERE pool.id IN (:pool_ids) " +
            "UNION " +
            "SELECT pool.derived_product_uuid FROM cp_pool pool WHERE pool.id IN (:pool_ids) " +
            "UNION " +
            "SELECT pp.product_uuid FROM cp2_pool_provided_products pp WHERE pp.pool_id IN (:pool_ids) " +
            "UNION " +
            "SELECT dpp.product_uuid FROM cp2_pool_derprov_products dpp WHERE dpp.pool_id IN (:pool_ids) ";

        Query prodQuery = entityManager.createNativeQuery(sql);
        Set<String> prodIds = new HashSet<>();

        int blockSize = Math.min(this.getInBlockSize(), this.getQueryParameterLimit() / 4);

        for (List<String> block : Iterables.partition(poolIds, blockSize)) {
            prodIds.addAll(prodQuery.setParameter("pool_ids", block).getResultList());
        }

        // Use the product UUIDs to select all related enabled content
        jpql = "SELECT DISTINCT pc.content, pc.enabled FROM ProductContent pc " +
            "WHERE pc.product.uuid IN (:product_uuids)";

        Query contentQuery = entityManager.createQuery(jpql);

        HashMap<Content, Boolean> activeContent = new HashMap<>();

        for (List<String> block : this.partition(prodIds)) {
            contentQuery.setParameter("product_uuids", block)
                .getResultList()
                .forEach(col -> {
                    Content content = (Content) ((Object[]) col)[0];
                    Boolean enabled = (Boolean) ((Object[]) col)[1];

                    activeContent.merge(content, enabled, (v1, v2) ->
                        v1 != null && v1.booleanValue() ? v1 : v2);
                });
        }

        return activeContent;
    }

    ...

    // ContentAccessManager.java
    this.ownerContentCurator.getActiveContentByOwner(owner.getId())
        .forEach((content, enabled) -> container.addContent(content, enabled));

```

This avoids the N+1 problem entirely, as we can verify by checking the output during the same spec test:

```
select pool0_.id as col_0_0_ from cp_pool pool0_ where pool0_.owner_id=? and pool0_.startDate<=? and pool0_.endDate>=?

SELECT pool.product_uuid FROM cp_pool pool WHERE pool.id IN (?, ?) UNION SELECT pool.derived_product_uuid FROM cp_pool pool WHERE pool.id IN (?, ?) UNION SELECT pp.product_uuid FROM cp2_pool_provided_products pp WHERE pp.pool_id IN (?, ?) UNION SELECT dpp.product_uuid FROM cp2_pool_derprov_products dpp WHERE dpp.pool_id IN (?, ?) 

select distinct productcon0_.content_uuid as col_0_0_, productcon0_.enabled as col_1_0_, content1_.uuid as uuid1_1_, content1_.created as created2_1_, content1_.updated as updated3_1_, content1_.arches as arches4_1_, content1_.contentUrl as contentU5_1_, content1_.entity_version as entity_v6_1_, content1_.gpgUrl as gpgUrl7_1_, content1_.content_id as content_8_1_, content1_.label as label9_1_, content1_.locked as locked10_1_, content1_.metadataExpire as metadat11_1_, content1_.name as name12_1_, content1_.releaseVer as release13_1_, content1_.requiredTags as require14_1_, content1_.type as type15_1_, content1_.vendor as vendor16_1_ from cp2_product_content productcon0_ inner join cp2_content content1_ on productcon0_.content_uuid=content1_.uuid where productcon0_.product_uuid in (? , ? , ?)
```

Note that this also keeps the logic for this problem in one place.

Reply:
This look very good and efficient. This certainty avoids the N+1 problem. 
thank for help me out with this.